Decoupling Core Topology Lookup from Hazelcast

jimwebber opened the pull request:
We now refresh our local view of topology on topology change,
when asked, we now return that view.

jimwebber committed:
Decoupling Core Topology Lookup from Hazelcast

We now refresh our local view of topology on topology change,
when asked, we now return that view.

apcj commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/discovery/HazelcastCoreTopologyService.java:
```diff
@@ -206,15 +212,17 @@ public void memberAdded( MembershipEvent membershipEvent )
             log.info( "Core member added %s", membershipEvent );
             log.info( "Current core topology is %s", coreServers() );
-            listenerService.notifyListeners( coreServers());
+            listenerService.notifyListeners( coreServers() );
```
Need to call refreshCoreTopology() before logging and notifying listeners, otherwise the log message will have the old topology, which will be very confusing. Maybe move both calls inside refreshCoreTopology() ?

jimwebber replied:
Okey.

apcj commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/discovery/HazelcastCoreTopologyService.java:
```diff
@@ -206,15 +212,17 @@ public void memberAdded( MembershipEvent membershipEvent )
             log.info( "Core member removed %s", membershipEvent );
             log.info( "Current core topology is %s", coreServers() );
-            listenerService.notifyListeners( coreServers());
+            listenerService.notifyListeners( coreServers() );
```
ditto

jimwebber replied:
Ditto, done.

apcj commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/discovery/HazelcastCoreTopologyService.java:
```diff
@@ -92,8 +92,10 @@ public void start()
-        listenerService.notifyListeners( coreServers());
+        mapRegistrationId = hazelcastInstance.getMap( EDGE_SERVER_BOLT_ADDRESS_MAP_NAME )
+                .addEntryListener( new OurEntryListener(), true );
+        listenerService.notifyListeners( coreServers() );
```
First time around coreServers() will return null, won't it?

jimwebber replied:
Re-ordered. Ta for that.

jimwebber committed:
More Decoupling Core Topology Lookup from Hazelcast

Re-ordered core topology lookups so that the locally cached copy
cannot be null after start.
Latest topology is gathered before listeners updated.
Moved logging into refresh method.
